### PR TITLE
build: add support for building for FreeBSD under Sourcehut

### PR DESCRIPTION
**Pull request by jszakmeister**

### CODE REVIEW

**Comment by jszakmeister** [comment]
@justinmk I think this is ready.  There are some test failures, but I believe they're legit vs. being environmental issues is Sourcehut.

**Comment by jszakmeister** [comment]
FYI: we could probably split things up a little more and have it do a basic build without tests, just as a data point though we'd still have to build everything again to do the unit and functional tests. *shrug*  I'm not sure what's best.  Also, I opted to put together a script to do the work since I wanted to build both debug and release and run tests for both (like we did under QuickBuild).  Sourcehut stops when a task fails, so this seemed like the only reasonable way to go about it.  We could take a different strategy and do something like have separate builds for debug and release... it would probably help the whole process go a bit faster.  Thoughts?

**Comment by justinmk** [comment]
> we could probably split things up a little more and have it do a basic build without tests, just as a data point though we'd still have to build everything again to do the unit and functional tests.

That's unfortunate :( Would be nice to have the different "stages", I wonder why sourcehut works like that.

**Comment by justinmk** [comment]
> > we could probably split things up a little more and have it do a basic build without tests, just as a data point though we'd still have to build everything again to do the unit and functional tests.

Maybe I misunderstood you, but the OpenBSD build has a separate "build" and "test" stage without needing to rebuild for the "test" stage. Example: https://builds.sr.ht/~jmk/job/86266

**Comment by jszakmeister** [comment]
> Maybe I misunderstood you, but the OpenBSD build has a separate "build" and "test" stage without needing to rebuild for the "test" stage. Example: https://builds.sr.ht/~jmk/job/86266

Yeah, you can do that, but once any task fails, the whole thing fails and sourcehut won't execute any more tasks.  So you can't try and do the other bits.  In this case, if we fail tests with the debug build, it would never try to do the release build and test.  Maybe it's what we want?  I personally like having it try to do everything that it can as it often sheds some light on issues.  I originally had it all broken out as separate tasks but didn't like that it didn't try the other tasks, so I went this direction.  I can put it back, if that's what you prefer.

**Comment by justinmk** [comment]
I see. For the common case, it's probably nicer for contributors if they are separated.

**Comment by blueyed** [comment]
As far as I can see build and test should be separate, but then there should be just 2 jobs maybe (debug build and release build).
OTOH we're not testing the release build explicitly on Travis either (but therefore it might be good to do so on Sourcehut then).

**Comment by justinmk** [comment]
I actually can't recall any time where the Release build caught anything not caught in the Debug build. I think the [nightly build](https://travis-ci.org/neovim/bot-ci) is good enough for that.

**Comment by jszakmeister** [comment]
> I actually can't recall any time where the Release build caught anything not caught in the Debug build. I think the [nightly build](https://travis-ci.org/neovim/bot-ci) is good enough for that.

There are some warnings in the compiler that won't actually get produced unless you have optimizations turned on (they require some of the optimization passes to happen to detect the issues they're looking for), so at the very least, it's good to do a release build.  I've also been on projects where a debug build passed a test but the release build did not.  Particularly, when it came to this bug in GCC (which still exists today): https://gcc.gnu.org/bugzilla/show_bug.cgi?id=18501

In this case, the compiler often assumes that an uninitialized variable is zero, and then proceeds to optimize things away, including branches that may have set it to something else (eek!).  It doesn't show up in debug builds (when optimizations are off), but does during a release build (when optimizations are on).

I'll split the tasks back up and let it fail early--it is nice to see from a user's point of view.  I have to run at the moment, but I'll get to it a little later today. :-)

**Comment by jszakmeister** [comment]
@justinmk I believe this is ready. :-)  Let me know if there are any more changes you'd like to see.

**Comment by blueyed** [comment]
@jszakmeister 
There are some test failures: https://builds.sr.ht/~jmk/job/86355#task-test-debug
One at least might be flaky (and fixed), so a rebase might help for that.

**Comment by justinmk** [comment]
The test failure seem related to localization(?)

**Comment by jszakmeister** [comment]
> The test failure seem related to localization(?)

One of them might have been related to that, but gettext is being installed.  It's not clear what is happening there.  The others seemed like the same sort of flaky stuff we've been fighting for a long time.

**Comment by jszakmeister** [comment]
> One at least might be flaky (and fixed), so a rebase might help for that.

Roger that.  Rebased and pushed up.  Let's see what happens. :-)

**Comment by blueyed** [comment]
```
[  FAILED  ] 1 test, listed below:
[  FAILED  ] test/functional/eval/let_spec.lua @ 61: :let multibyte env var to child process #8398 #9267
test/functional/eval/let_spec.lua:69: Expected objects to be the same.
Passed in:
(string) ''
Expected:
(string) 'AìaB'

stack traceback:
	test/functional/eval/let_spec.lua:69: in function <test/functional/eval/let_spec.lua:61>

[  ERROR   ] 1 error, listed below:
[  ERROR   ] test/functional/ui/messages_spec.lua @ 1050: ui/msg_puts_printf output multibyte characters correctly
test/functional/ui/screen.lua:587: Row 1 did not match.
Expected:
  |*^Exモードに入ります. ノー |
  |*マルモードに戻るには"visu|
  |*al"と入力してください.   |
  |:                        |
  |                         |
Actual:
  |*^Entering Ex mode.  Type "|
  |*visual" to go to Normal m|
  |*ode.                     |
  |:                        |
  |                         |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.  

stack traceback:
	test/functional/ui/screen.lua:587: in function '_wait'
	test/functional/ui/screen.lua:370: in function 'expect'
	test/functional/ui/messages_spec.lua:1082: in function <test/functional/ui/messages_spec.lua:1050>


 18 SKIPPED TESTS
 1 FAILED TEST
 1 ERROR
```

Do they work locally for you on FreeBSD?

btw: you can use `[ci skip]` to skip builds on Travis / AppVeyor (currently ignored by Sourcehut (https://todo.sr.ht/%7Esircmpwn/builds.sr.ht/161)).

**Comment by blueyed** [comment]
Well, at least the second one appears to be skipped on CI, but that does not match Sourcehut?!
https://github.com/neovim/neovim/blob/0571145c40b0a2ae106acc43891c2680c76b8383/test/functional/ui/messages_spec.lua#L1070-L1074
https://github.com/neovim/neovim/blob/0571145c40b0a2ae106acc43891c2680c76b8383/test/helpers.lua#L716-L723

File: .builds/freebsd.yml [hunk 1]
```diff
@@ -16,22 +16,24 @@ sources:
...
+#     mkdir -p neovim/build/release
+#     cd neovim/build/release
+#     cmake -G Ninja -DTRAVIS_CI_BUILD=ON -DCMAKE_BUILD_TYPE=Release -DMIN_LOG_LEVEL=3 ../..
+#     ninja -v nvim unittest-prereqs functionaltest-prereqs
+# - test-debug: |
+#     cd neovim/build/debug
+#     ninja -v unittest functionaltest
+# - test-release: |
+#     cd neovim/build/release
+#     ninja -v unittest functionaltest
```
Comment (blueyed) [inline .builds/freebsd.yml]: @jszakmeister please use `[ci skip]` in the commit message's subject to skip Travis/AppVeyor builds.

Comment (jszakmeister) [reply .builds/freebsd.yml]: @blueyed Ah, it sounded like it could be at the end of the commit message. :-(

**Comment by blueyed** [comment]
Looking great!

Can we run `build-release` and `test-release` only for tags maybe?

And `make oldtest` should/could be included.

**Comment by blueyed** [comment]
Failure is a flaky / strange one: https://github.com/neovim/neovim/pull/11086#issuecomment-535247575

**Comment by jszakmeister** [comment]
> Looking great!

Thanks!

> Can we run `build-release` and `test-release` only for tags maybe?

To be honest, I'd rather have it run release builds if for no other reason than some warnings/errors don't happen without them.  That said, you guys deal with stuff way more often than me, so it's your call.

Outside of that, I'm not sure how we test for that... I suppose we could use `git describe` or something similar to try and figure that out?  That said, I'm not sure how sourcehut is invoked in that situation either since tagging a release isn't necessarily happening through a pull request.  I think I'd rather just drop one set of tasks for now.

> And `make oldtest` should/could be included.

I'll have to rework how I build then.  Right now, I'm using straight up CMake and Ninja but going through the Makefile to build oldtest will require things that only the Makefile is doing.  Shouldn't be too hard to fix it up though.

**Comment by blueyed** [comment]
> I'd rather have it run release builds

Ok, but then do not use the test build jobs?  :+1:

>> `make oldtest`

> I'll have to rework how I build then. Right now, I'm using straight up CMake and Ninja but going through the Makefile to build oldtest will require things that only the Makefile is doing. Shouldn't be too hard to fix it up though.

Given that you use `.deps` and `build` dirs it should just work, no?  (also not optimal, e.g. since `build/.ran-cmake` is not there yet)

Maybe using `make cmake` instead of `cmake` directly is better in general?

**Comment by jszakmeister** [comment]
> Ok, but then do not use the test build jobs? 👍

You don't want to run tests on the release builds?

> Given that you use `.deps` and `build` dirs it should just work, no? (also not optimal, e.g. since `build/.ran-cmake` is not there yet)

Maybe... it wasn't too hard to convert though. :-)  Running through some checks right now.

File: .builds/freebsd.yml [hunk 2]
```diff
@@ -19,26 +19,36 @@ sources:
...
-    cd neovim/build
-    cmake -G Ninja -DTRAVIS_CI_BUILD=ON -DCMAKE_BUILD_TYPE=Release -DMIN_LOG_LEVEL=3 ..
-    ninja -v nvim unittest-prereqs functionaltest-prereqs
+    cd neovim
+    gmake CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="${CMAKE_EXTRA_FLAGS}" nvim
 - test-release: |
-    cd neovim/build
-    ninja -v unittest functionaltest
+    cd neovim
+    gmake CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="${CMAKE_EXTRA_FLAGS}" unittest functionaltest
```
Comment (blueyed) [inline .builds/freebsd.yml]: No need to pass cmake stuff here.

File: .builds/freebsd.yml [hunk 3]
```diff
@@ -19,26 +19,36 @@ sources:
...
+- test-oldtest: |
+    cd neovim
+    gmake CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="${CMAKE_EXTRA_FLAGS}" oldtest
```
Comment (blueyed) [inline .builds/freebsd.yml]: same here.

**Comment by blueyed** [comment]
> > Ok, but then do not use the test build jobs? +1
> 
> You don't want to run tests on the release builds?

No, I've meant to only build/test the Release type, but not build/test the Debug type.

File: .builds/freebsd.yml (hunk 2 above)
Comment (jszakmeister) [reply .builds/freebsd.yml]: Yeah... copy and past for the moment.  I'll clean it up before calling it done.

**Comment by jszakmeister** [comment]
> No, I've meant to only build/test the Release type, but not build/test the Debug type.

Roger that.  That's exactly what I was thinking.

**Comment by jszakmeister** [comment]
> And `make oldtest` should/could be included.

This looks to be causing a weird failure on sourcehut.  It dies and then says:
```text
Error fetching logs for task "test-oldtest"
```

Running locally on my FreeBSD system, I don't see that kind of failure on mv VM--but it's also fairly old.

**Comment by jszakmeister** [comment]
I tried building and running oldtest as a debug build on sourcehut too, but that didn't work either.  So the oldtest stuff is disabled for now.  I'm not sure if it helps or not, but locally I see this:

```text
From test_cscope.vim:
Found errors in Test_cscopeWithCscopeConnections():
function RunTheTest[37]..Test_cscopeWithCscopeConnections line 32: Pattern '\\n(1 of 42): <<mf_hash_init>> mf_hash_init(&ht);' does not match '\n(1 of 19): <<mf_hash_init>> mf_hash_init(&ht);'
function RunTheTest[37]..Test_cscopeWithCscopeConnections line 32: Pattern '\\n(1 of 42): <<mf_hash_init>> mf_hash_init(&ht);' does not match '\n(1 of 19): <<mf_hash_init>> mf_hash_init(&ht);'
function RunTheTest[37]..Test_cscopeWithCscopeConnections line 88: Expected '(1 of 1): <<global>> #include <assert.h>' but got '(1 of 2): <<global>> #include <assert.h>'

From test_alot.vim:
Found errors in Test_libcall_libcallnr():
Caught exception in Test_libcall_libcallnr(): Vim(call):dlerror = "/usr/local/lib/: read error: Invalid argument" @ function RunTheTest[37]..Test_libcall_libcallnr, line 32
/usr/home/quickbuild/projects/neovim/build/bin/nvim -u unix.vim -U NONE -i NONE --noplugin --headless -u NONE -S summarize.vim messages
```

I think it should be ready now.  Let me know if you need anything else, but it might take me 2 or 3 days to get back to it... lots going on this week.

**Comment by blueyed** [comment]
Can you provide a link to a / the last failing build for oldtest for reference, please?

File: .builds/freebsd.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    gmake deps
+- build: |
+    cd neovim
+    gmake CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="${CMAKE_EXTRA_FLAGS}" nvim
+- test: |
+    cd neovim
+    gmake unittest functionaltest
+
+# Unfortunately, oldtest is tanking hard on sourcehut's FreeBSD instance
+# and not producing any logs as a result. So don't do this task for now.
```
Comment (blueyed) [inline .builds/freebsd.yml]: ah, apparently no logs..

**Review by blueyed — APPROVED** [review]

**Review by justinmk — APPROVED** [review]

**Comment by jszakmeister** [comment]
> Can you provide a link to a / the last failing build for oldtest for reference, please?

Yep.  Here it is: https://builds.sr.ht/~jmk/job/91973

I'm not sure how long those builds hang around.

**Pull request merged**